staging/src/k8s.io/apimachinery/pkg/runtime/extension.go

```diff
@@ -18,16 +18,48 @@ package runtime
 
 import (
 	"bytes"
-	"encoding/json"
 	"errors"
+
+	"k8s.io/apimachinery/pkg/util/json"
 )
 
+func (re RawExtension) ToUnstructured() interface{} {
+	// This conversion should be possible when Object is non-nil, or when ContentType is
+	// recognized and matches the contents of Raw. It's possible to have a RawExtension with an
+	// unrecognized ContentType, or Raw bytes that are invalid for a recognized ContentType
+	// (either by constructing the RawExtension directly in code or by unmarshaling a
+	// RawExtension from Protobuf). Is that a bug?
```

Comment:
> It's possible to have a RawExtension with an
> unrecognized ContentType

Yes, we should be validating at system boundries, but unstructured can be use in a wide range of places (on the client and in servers) where we shouldn't assume the data is validated.  I would, however, consider such data malformed, and handling of it can reflect that.  I'd expect errors where possible (as opposed to silently handling the malformed data in potentially bad ways)

> or Raw bytes that are invalid for a recognized ContentType

Also yes, for the same reasons, but again, the RawExtension would be malformed.  Favoring errors here too.

> (either by constructing the RawExtension directly in code or by unmarshaling a
> RawExtension from Protobuf). Is that a bug?

I was going to say, "yes this is a bug if it happens within the kube-apiserver", but I think we've got the legacy case of non-json bytes being written to RawExtension data when protobuf is used that we've got to deal with, so there might be a class of malformed data cases that can happen from that case that are not really a kube-apiserver bug?

Reply:
I avoided implementing UnstructuredConverter on RawExtension for now. If conversion to unstructured passes instead through the json.Marshaler implementation, and the raw bytes can't be automatically transcoded to JSON, the call to MarshalJSON will return an error.